plugins/plugin-factory-github/codenvy-plugin-factory-github-server/src/main/java/com/codenvy/plugin/factory/github/server/GithubFactoryParametersResolver.java

```diff
@@ -0,0 +1,194 @@
+                    "The given github url %s is not a valid URL github url. It should start with https://github.com/<user>/<repo>",
+                    githubUrl));
+        }
+
+        // ok so now get matching group
+        int regexpGroupCount = matcher.groupCount();
+        if (regexpGroupCount < 2) {
+            throw new BadRequestException(String.format(
+                    "The given github url %s is not a valid URL github url. It should start with https://github.com/<user>/<repo>",
+                    githubUrl));
+        }
+
+        String repoUser = matcher.group(1);
+        String repoName = matcher.group(2);
+        String branchName = null;
+        String subfolder = null;
+        if (regexpGroupCount == 4) {
+            branchName = matcher.group(3);
+            subfolder = matcher.group(4);
+        }
+
+        // build factory for the given github url
+        Map<String, String> parameters = new HashMap<>();
+        if (branchName != null) {
+            parameters.put("branch", branchName);
+        }
+        if (subfolder != null && !"/".equals(subfolder)) {
+            parameters.put("keepDir", subfolder);
+        }
+
+        // github location for the clone
+        String githubLocation = "https://github.com/".concat(repoUser).concat("/").concat(repoName);
+
+
+        // default
+        final String DEFAULT_DOCKER_FILE_LOCATION = "https://dockerfiles.codenvycorp.com/templates-4.0/factory/factory-dockerfile";
+
+        // use master if not specified
+        if (branchName == null) {
+            branchName = "master";
+        }
+
+        // check if repository contains a codenvy Dockerfile
+        String dockerFileLocation = null;
+        String githubDockerFileLocation =
+                "https://raw.githubusercontent.com/".concat(repoUser).concat("/").concat(repoName).concat("/").concat(branchName)
+                                                    .concat("/.codenvy.dockerfile");
+        HttpURLConnection httpURLConnection = null;
+        try {
+            httpURLConnection = (HttpURLConnection)new URL(githubDockerFileLocation).openConnection();
+            httpURLConnection.setConnectTimeout(CONNECTION_TIMEOUT);
+            final int responseCode = httpURLConnection.getResponseCode();
+            if (responseCode == HttpURLConnection.HTTP_OK) {
+                dockerFileLocation = githubDockerFileLocation;
+            }
+        } catch (IOException ioe) {
+            LOG.debug("Unable to check if remote location {0} is available or not", githubDockerFileLocation);
+        } finally {
+            if (httpURLConnection != null) {
+                httpURLConnection.disconnect();
+            }
+        }
+
+        // apply default location if there is none
+        if (dockerFileLocation == null) {
+            dockerFileLocation = DEFAULT_DOCKER_FILE_LOCATION;
+        }
+
+        String envName = repoUser;
+
+        SourceStorageDto
```

Comment:
This method is so big. I would like you split it out. But AFAIK @skabashnyuk doesn't like private methods with single usage. 
@skabashnyuk @benoitf  WDYT about splitting this method?